### DISCUSSION TITLE
layout="fill" doesn't respect maximum dimensions & name is poorly worded

### DISCUSSION THREAD

**Discussion by agcty** [vercel/next.js#20523] [Help]
## Problem
When using the next/image component with the layout="fill" option, it doesn't behave like the native img tag. This makes it unsuitable for using the image component in scenarios where you want to define maximum dimensions the image can fill. Also, in my opinion, "fill" is poorly worded and is highly misleading.

Take this example:
```jsx
import Image from "next/image";


function ImageTest() {
  return (
    <div>
      {/* this will render a 0x0 px image */}
      <div
        className="test"
        style={{
          position: "relative",
          maxHeight: "200px",
          maxWidth: "200px",
        }}
      >
        <Image
          src="https://images.unsplash.com/photo-1609050470947-f35aa6071497?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=3450&q=80"
          layout="fill"
        />
      </div>

      {/* this works like it should -> image is portrait so it will fill the full 500px width and fill the height automatically */}
      <div
        style={{
          position: "relative",
          marginTop: "50px",
          maxHeight: "200px",
          maxWidth: "500px",
        }}
      >
        <img src="https://images.unsplash.com/photo-1609050470947-f35aa6071497?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=3450&q=80" />
      </div>
    </div>
  );
}

export default ImageTest;
```

The above code will render:
![image](https://user-images.githubusercontent.com/12101091/103180218-07e8f780-4894-11eb-85b1-fa58522479d6.png)

Note: There should be two images visible but only the native img tag works with the styling provided.

## Consequences
This makes the next/image component unsuitable for about 70% (wild guess but feels right) of image scenarios and I'm really curious why people are not annoyed by it. Is there something I'm missing? Some secret styling rules? Or do people just not use next/image and nobody cares? Would love to know.

What makes it unsuitable for most scenarios is that most of the times, you don't know what the actual image size is and what aspect ratio you need to render at. I know that for performance reasons you should always provide width and height but most of the times this doesn't make sense at all. You just want to be able to say: Here's a container with the maximum dimensions of 500px width & 400px height - figure out how to fit in there while preserving the aspect ratio.

The wording layout="fill" is misleading because it only fills a container with a predefined size, NOT however a container with maximum dimensions.

## Solution
To solve this issue, remove the styling from the parent div tag and some styling from the img tag that is being rendered by next/image and it behaves as it should.

My proposal is to add a new layout mode called "none" that removes the styling altogether and leaves the image component with lazy loading etc.

Take a look here at what removing the styling does:
https://www.loom.com/share/18f80ddc63f14892809f1abeeca2c7df

@luknl @nastari @apotdevin @trompx @iampava

**Comment by agcty** [comment]
Has anyone solved this issue in any other way that doesn't involve overwriting styles?

**Reply by Timer** [reply]
Here's two options:
```tsx
import Image from "next/image";

export default function IndexPage() {
  return (
    <main>
      <div
        style={{
          position: "relative",
          height: "200px",
          width: "200px"
        }}
      >
        <Image
          src="https://images.unsplash.com/photo-1609050470947-f35aa6071497?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=3450&q=80"
          layout="fill"
          objectFit="contain"
        />
      </div>
      <Image
        src="https://images.unsplash.com/photo-1609050470947-f35aa6071497?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=3450&q=80"
        width={200}
        height={200}
        objectFit="contain"
      />
    </main>
  );
}
```

![image](https://user-images.githubusercontent.com/616428/103453774-61459200-4cab-11eb-8f9f-ff5c92c6ee31.png)

**Comment by cballenar** [comment]
@agcty were you able to find any solutions to this?

I'm struggling with the same issue due to variable image ratios. I have a fixed width but the images expected are of variable sizes and ratios.

So far wherever I look it seems to always require a manual specification of the width AND height with no other workaround than a hack that relies on css with `!important`s.
